omegaconf/basecontainer.py

```diff
@@ -55,25 +55,24 @@ def is_mandatory_missing(val: Any) -> bool:
 
         val = _get_value(value)
         has_default = default_value is not DEFAULT_VALUE_MARKER
-        if has_default and (val is None or is_mandatory_missing(val)):
+
+        if has_default and val is None:
             return default_value
 
+        if is_mandatory_missing(val):
+            if has_default:
+                return default_value
+            raise MissingMandatoryValue("Missing mandatory value: $FULL_KEY")
+
         resolved = self._maybe_resolve_interpolation(
             parent=self,
             key=key,
             value=value,
-            throw_on_missing=not has_default,
             throw_on_resolution_failure=not has_default,
         )
         if resolved is None and has_default:
```

Comment:
A bit uneasy here about confusing a legit None result with a None because of an exception.
can we try catch here instead and handle it by returning the default if has_default or letting it fly otherwise?

Reply:
Not sure what you mean by "legit None". Getting a None means that something went wrong (it was also the case previously). Otherwise we are supposed to get a `Node` (that may wrap a None value).

Reply:
```
foo : null
bar: ${foo}
```

bar is a "legit None".

Looking at the code it was not clear to me that resolved is a node.
rename to resolved_node ?

Reply:
Ok, yes in that case `resolved` would still be a node. I renamed the variable to `resolved_node` in e2d3e3b